experimental/Schemes/elliptic_surface.jl

```diff
@@ -0,0 +1,1116 @@
+  # and for the degree of c to be bounded above by
+  # k + 2*OP + 4 + degree(xd)
+  eq1 = collect(coefficients(r))
+  eq2 = [coeff(c,i) for i in (k + 2*OP + 4 + degree(xd) + 1):degree(c)]
+  eqns = vcat(eq1, eq2)
+
+  # collect the equations as a matrix
+  cc = [[coeff(j, abi) for abi in ab] for j in eqns]
+  M = matrix(B, length(eqns), length(ab), reduce(vcat,cc, init=elem_type(base)[]))
+  # @assert M == matrix(base, cc) # does not work if length(eqns)==0
+  kerdim, K = kernel(M)
+  result = Tuple{elem_type(Bt),elem_type(Bt)}[]
+  t = gen(Bt)
+  for j in 1:kerdim
+    aa = reduce(+, (K[i+1,j]*t^i for i in 0:dega), init=zero(Bt))
+    bb = reduce(+, (K[dega+i+2,j]*t^i for i in 0:degb), init=zero(Bt))
+    push!(result, (aa, bb))
+  end
+  # confirm the computation
+  @assert kerdim == 2*k + OP # prediced by Riemann-Roch
+  for (a,b) in result
+    @assert mod(a*xn - b*yn, xd) == 0
+    @assert degree(a) <= k + 2*OP
+    @assert degree(b) <= k + 2*OP - 2 - 1//2*degree(xd)
+    @assert degree(a*xn - b*yn) <= k + 2*OP + 4 + degree(xd)
+  end
+  return result
+end
+
+@doc raw"""
+    linear_system(X::EllipticSurface, P::EllCrvPt, k::Int64) -> LinearSystem
+
+Compute the linear system ``|O + P + k F|`` on the elliptic surface ``X``.
+Here ``F`` is the class of the fiber over ``[0:1]``, ``O`` the zero section
+and ``P`` any section given as a point on the generic fiber.
+"""
+function linear_system(X::EllipticSurface, P::EllCrvPt, k::Int64)
+  X.bundle_number == 2 || error("linear system implemented only for elliptic K3s")
+  FS = function_field(weierstrass_model(X)[1])
+  U = weierstrass_chart(X)
+  (x,y,t) = ambient_coordinates(U)
+
+  sections = elem_type(FS)[]
+  if iszero(P[3])
+    append!(sections, [FS(t)^(i-k) for i in 0:k])
+    append!(sections, [FS(t)^(i-k)*FS(x) for i in 0:k-4])
+  elseif iszero((2*P)[3])
+    error("not implemented for 2-torsion sections")
+  else
+    xn = numerator(P[1])
+    xd = denominator(P[1])
+    yn = numerator(P[2])
+    yd = denominator(P[2])
+
+    I = ambient_closure_ideal(U)
+    IP = ideal([x*xd(t)-xn(t),y*yd(t)-yn(t)])
+    issubset(I, IP) || error("P does not define a point on the Weierstrasschart")
+
+    @assert gcd(xn, xd)==1
+    @assert gcd(yn, yd)==1
+    ab = _prop217(generic_fiber(X), P, k)
+    d = divexact(yd, xd)(t)
+    den = t^k*(x*xd(t) - xn(t))
+    for (a,b) in ab
+      c = divexact(b*yn - a*xn, xd)
+      num = a(t)*x+b(t)*d*y + c(t)
+      push!(sections, FS(num//den))
+    end
+  end
+  return sections
+  sectionsX = pullback(X.blowup).(sections)
+  return sectionX
+end
+
+"""
+  elliptic_parameter(X::EllipticSurface, F::ZZMat) -> LinearSystem
+
+Return the complete linear system ``|F|``.
+
+Here `F` must be given with respect to the basis of
+`algebraic_lattice(X)` and be an isotropic nef divisor.
+"""
+function elliptic_parameter(X::EllipticSurface, F::ZZMatrix)
+  NS = algebraic_lattice(X)
+  @req F*gram_matrix(NS)*transpose(F) "not an isotropic divisor"
```

Comment:
What is the requirement here? That it's zero, right? 

What is missing for the implementation below to work? Anything on my side?

Reply:
Yes zero. I think we can implement it now.

Reply:
It is a matter of automatizing the bookkeeping.

Reply:
Done.